feat(common): add <link> preload tag on server for priority img

yharaskrik opened the pull request:
While in Angular Universal, for images that are priority add a preload <link> tag to the <head> to ensure the image is preloaded before it is rendered. This resolves a warning when running Lighthouse.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Currently when using the `NgOptimizedImage` directive on `<img>` tags with the `priority` attribute (as suggested in the docs), Lighthouse will show a warning about preloading some assets to improve LCP metrics.

More here: https://web.dev/preload-critical-assets/#effects-of-preloading-on-core-web-vitals

Issue Number: N/A


## What is the new behavior?

While rendering a page in Angular Universal if an image is going to be loaded with `priority="true"` then add a preload <link> tag to the <head> so that when the browsers renders the static HTML (rendered from the server) there are preload links for the browser to start loading the LCP images ahead of time and thus satisfying Lighthouse.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Investigation numbers here: https://twitter.com/JayCooperBell/status/1566231614849331200

yharaskrik committed:
feat(common): add <link> preload tag on server for priority img

This commit adds a logic that generates preload tags for priority images, when rendering happens on the server (e.g. Angular Universal).

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -358,6 +368,16 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+
+  private createPreloadLinkTag(url: string): void {
+    const preload = this.document.createElement('link');
+    preload.setAttribute('fetchpriority', 'high');
```
Fetchpriority shouldn't be needed here (already added to the <img> tag itself).

yharaskrik replied:
Does `fetchpriority` on the `img` tag affect the `fetchpriority` of the `link` tag then?

kara replied:
For more context: adding high fetchpriority on the preload tag for an image only makes sense if 1) you're preloading a background image defined in the CSS (since there's no way to define fetchpriority in CSS) or 2) if you're doing something fancy with only preloading certain images in a srcset (using media queries, etc).

For 1), we know this isn't happening because this preload tag is added by the image directive on an image tag. For 2), this might be something we can start supporting as part of auto-generating srcsets, but too early to implement that as part of this PR.

The best practice is to avoid adding high fetchpriority in the general case because it would cause the preload to jump in front of other scripts that could be more critical. So adding it to the `<img>` tag should be sufficient for our purposes here.

yharaskrik replied:
Amazing thank you for the explanation! Makes perfect sense. 

Commit incoming with a test!

kara reviewed (changes requested):
Thanks for the PR! In general, looks good, but needs a test

yharaskrik commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -301,10 +305,16 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
       this.setHostAttribute('srcset', this.getRewrittenSrcset());
     }
+
+    if (this.isServer && this.priority) {
```
@kara One question I was thinking about, if the same image is loaded multiple times on a way then this will create a preload for that every time, should I add a 

```
document.querySelector(`link[href=${url}]`)
```
 so that we are not adding double preload tags or is that not going to be a problem and we might as well just save the extra DOM querying?

AndrewKushnir replied:
Good catch 👍 There is one more situation that we've discussed internally: there might be a situation when a `priority` is added (potentially by accident) to an `<img>` element in an `*ngFor` which iterates over a large set of images. As a result, we'd end up generating a large set of `<link>`s, requesting a preload for all of them.

A possible solution might be to introduce a Set, which would contain all images for which we generated a `<link>`. This can help us to:
- make sure we don't generate the same `<link>` twice
- keep track of the number of images for which a `<link>` tag was requested and throw an error if it's larger than N (N might be somewhere between 3-10?)

We can refer to that set via an InjectionToken provided in `root` (so we count the number of images per application) and wrap it into an `ngDevMode` check (to tree-shake it away in prod mode).

Here is how the token may look like:
```
// ... docs ...
const PRELOADED_IMAGES = new InjectionToken<Set<string>>('PRELOADED_IMAGES', {
  providedIn: 'root',
  factory: () => new Set(),
});

// ... docs ...
const PRELOADED_IMAGES_CAP = 5;
```

and in the directive code you can request it as:

```
  private createPreloadLinkTag(url: string): void {
    if (ngDevMode) {
      const preloadedImages = this.injector.get(PRELOADED_IMAGES);
      if (preloadedImages.size > PRELOADED_IMAGES_CAP) {
        throw new RuntimeError( ... );
      }

      // Avoid generating the same `<link>` tag.
      if (preloadedImages.has(url)) return;

      preloadedImages.add(url);
    }
    const preload = this.document.createElement('link');
    preload.setAttribute('as', 'image');
    preload.href = url;
    preload.rel = 'preload';

    this.document.head.appendChild(preload);
  }
```

Please let me know if any additional information is required, would be happy to help.

Thank you.

yharaskrik replied:
thanks for your response Andrew, but I am a bit confused here, why would we want to treeshake the de-dup logic in prod mode?

I can see why we would want to de-dup the runtime error but the actual set checking should be done throughout should it not? So something like this:

```
private createPreloadLinkTag(url: string): void {
    const preloadedImages = this.injector.get(PRELOADED_IMAGES);
    
    if (ngDevMode) {  
      if (preloadedImages.size > PRELOADED_IMAGES_CAP) {
        throw new RuntimeError( ... );
      }
    }

      // Avoid generating the same `<link>` tag.
      if (preloadedImages.has(url)) return;

      preloadedImages.add(url);

    const preload = this.document.createElement('link');
    preload.setAttribute('as', 'image');
    preload.href = url;
    preload.rel = 'preload';

    this.document.head.appendChild(preload);
  }
```

If we did it the way that you proposed we could end up duplicating `<link>` tags in the `<head>` while in prod mode (which we don't want?), we always want to de-dup the preload links right but want to prevent too many unique links being added?

AndrewKushnir replied:
Yeah, you are right, the dedup logic should be outside of the `ngDevMode` check 👍

yharaskrik replied:
OK sounds good! What would we like the max number of preloads to be?

AndrewKushnir replied:
> What would we like the max number of preloads to be?

We should probably make it configurable with a reasonable default (I'd vote for 5). May be something like this:

```
const DEFAULT_PRELOADED_IMAGES_LIMIT = 5;

/**
 * ... docs, example usages ...
 *
 * @publicApi
 * @developerPreview
 */
export const PRELOADED_IMAGES_LIMIT =
    new InjectionToken<number>('PRELOADED_IMAGES_LIMIT', {
      providedIn: 'root',
      factory: () => DEFAULT_PRELOADED_IMAGES_LIMIT,
    });
```

In this case we can also include a note into the error message that developers can increase the limit by providing the necessary value via `PRELOADED_IMAGES_LIMIT` token like `{provide: PRELOADED_IMAGES_LIMIT, useValue: 100}`. 

@kara WDYT?

AndrewKushnir replied:
This also reminds me of this thread: https://github.com/angular/angular/pull/47157#discussion_r969013840.

We should probably consolidate all config params into a single injection token like this:

```
export interface NgOptimizedImageConfig {
  maxPreloadedImages: number;

  // This replaces the `PRECONNECT_CHECK_BLOCKLIST` token from the public API.
  preconnectCheckBlocklist: string[];
}

export const NG_OPTIMIZED_IMAGE_CONFIG = new InjectionToken<NgOptimizedImageConfig>('NG_OPTIMIZED_IMAGE_CONFIG');
```

We'll discuss the config question more and I'll get back with additional information.

yharaskrik replied:
Sounds good! I'll hold off and the injection token for now!

yharaskrik replied:
Definitely agree though

yharaskrik replied:
missing comments but I implemented the injection tokens and stuff

AndrewKushnir replied:
@yharaskrik thanks for the update!

I was also playing with the directive config and realized that we should remove `PRECONNECT_CHECK_BLOCKLIST` token from the public API (in favor of the new config token). Also I found a potential issue where the `PreconnectLinkChecker` class was using stale info, ignoring extra configuration that might happen on a component level (in addition to setting up providers during the bootstrap). I've started making some changes here: https://github.com/angular/angular/commit/78e761eb23b08660536571e38a2d0d93eeb9e5f2. If everything goes well with tests, I should have a PR out within the next day or so and once it's merged (hopefully soon as well), we can rebase your PR to take the advantage of consolidated config.

I will keep this thread updated. Once again, thanks for your help with this change @yharaskrik 👍

yharaskrik replied:
Yes we definitely should and instead merge them together into a single config, I investigated doing that but realized that it would expand the scope of this PR unnecessarily. Once you get that switched up then I will make the necessary changes to this one!

https://media.giphy.com/media/Hk8vmotu0aQ6x2a21b/giphy.gif

kara replied:
5 sounds good for a reasonable default, but I don't think we should make it configurable. Preloading 10 images for example is a huge performance bottleneck and we shouldn't imply that's an OK thing to do

yharaskrik replied:
@kara what if we left it configurable but if they override it we print a warning telling them its probably not a good idea and wrap that in an `if (ngDevMode)` to tree shake it in prod?

kara replied:
I'd argue for making it not configurable, at least at first. This is not a feature we could remove easily once we add it :) If we see there's actually demand for configuring preload count (and good reasons!), we can always add the feature later.

yharaskrik replied:
Fair enough!

AndrewKushnir replied:
Just wanted to mentioned that I'm ok to avoid exposing a config option for this. As Kara mentioned, we can introduce a config option to either change the limit or disable this feature later if needed, based on the feedback from the community and specific use-cases.

@yharaskrik in this case, this PR is fully unblocked from any config-related changes that I mentioned above, so we can proceed with further updates and review.

yharaskrik replied:
I think I will just need to remove the configuration for it then and leave in the default as the hard limit.

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -381,6 +394,32 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+
+    const preload = this.document.createElement('link');
+    preload.setAttribute('as', 'image');
+    preload.href = url;
```
Ohh just realized that the tag doesn't currently support responsive images. We'll want to use `imagesrcset` (from `this.getRewrittenSrcset()`) and `imagesizes` here (from `sizes`) to make sure we're preloading the right width image if `srcset` is set.

yharaskrik replied:
Nice catch!

yharaskrik replied:
So unless I'm really missing it I don't see a this.sizes, should I parse (and cache) the properly formatted image URL in the function that parses the ngSrcset property so that the URLs can be referenced later (like when adding the link tag)? Then I can add the right image to the head if I can know what size will actually be rendered. Thanks Kara!

yharaskrik replied:
Ok, when I get back to my computer I will take a look into this and see what I can come up with!

AndrewKushnir commented:
@yharaskrik thanks again for working on this feature! We think it'd bring a lot of value in SSR scenarios and we'd like to merge it before the v15 feature freeze date (Oct 12). Please let us know if you may have a chance to look into this and/or you want us to help with some of the changes. If I understand correctly, the remaining items are:
- addressing the feedback from Kara regarding the `srcset` support, see https://github.com/angular/angular/pull/47343#pullrequestreview-1114128028
- drop the config-related changes (just use 5 as the default)

After that we can start the final code-review round, perform the necessary tests in Google's codebase and we should be ready for merge.

Thank you.

yharaskrik commented:
@AndrewKushnir sounds good! Just replied to Kara for some additional clarification as unless I'm missing something I don't see what she is referring to currently in the code, but maybe I'm missing it!

I'll make the config related changes tonight hopefully!

AndrewKushnir commented:
@yharaskrik just wanted to ask if you want us to help somehow with the changes from this PR and/or there are open questions we can help address.

yharaskrik commented:
> @yharaskrik just wanted to ask if you want us to help somehow with the changes from this PR and/or there are open questions we can help address.

Hey @AndrewKushnir! Sorry things got crazy this last week, I do have a question about @kara comment that I have been meaning to post, Kara you are totally right that is is preloading the wrong one, but I am not following how I am going to figure out what one to preload.

We can get the `sizes` and `srcsest` as an `@Input` as you described (we are already reading srcset), but that doesn't tell me what one the user will be loading. I am assuming we don't want to prelaod every size (kinda defeats the purpose) so how should I determine what size that the user is going to be loading, wouldn't it depend on the screensize and what srcset gets matched? Am I misunderstanding the `sizes` attribute?

Let me know what I am missing here thanks you two!

kara commented:
@yharaskrik Oh, the whole point of `imagesrcset` and `imagesizes` is that you *don't* have to know which one is being loaded. The browser will determine that. `imagesrcset` should have the same value as `srcset`. `imagesizes` should have the same value as `sizes`. Just copy them over and the browser will preload the correct URL.

There's more context in the docs here:
https://web.dev/preload-responsive-images/#imagesrcset-and-imagesizes

yharaskrik commented:
> @yharaskrik Oh, the whole point of `imagesrcset` and `imagesizes` is that you _don't_ have to know which size is being loaded. The browser will determine that. `imagesrcset` should have the same value as `srcset`. `imagesizes` should have the same value as `sizes`. Just copy them over and the browser will preload the correct URL.
> 
> There's more context in the docs here: https://web.dev/preload-responsive-images/#imagesrcset-and-imagesizes

OH those attributes are on the `<link>` tag! 🤦 That is why I was so confused.

yharaskrik commented:
Ok there we go, got it done! Sorry! That was way easier than I thought it was. I changed the prelaod link tags `href` to just the plain `ngSrc` input as that is what it seemed like it was supposed to be based on the link that @kara sent.

AndrewKushnir commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -8,16 +8,102 @@
 import {PRECONNECT_CHECK_BLOCKLIST} from '../../src/directives/ng_optimized_image/preconnect_link_checker';
 
 describe('Image directive', () => {
+  it('should create a preconnect `<link>` element in the `<head>` for the image when priority is true',
```
I think it'd be great to have a `describe` block here:
```
describe('preconnect <link> element on a server', () => ...);
```

AndrewKushnir commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -8,16 +8,102 @@
 
 describe('Image directive', () => {
+  it('should create a preconnect `<link>` element in the `<head>` for the image when priority is true',
+     () => {
```
Could you please also add a test to exercise a code path that throws an error when a number of priority images is more than 5 (default)? I think it'd be also great to have a similar test, but without `PLATFORM_SERVER_ID`, to verify that there are no limitations when an app is rendered in a browser.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -285,6 +291,13 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+  /**
+   * Value of the `sizes` attribute if set on the host `<img>` element.
+   * This input is exclusively read to pass along to the preload `<link>` attribute for this image.
+   * @internal
```
We should probably use `@nodoc` here to avoid exposing an input in the docs, but still keep it in DTS files.
```suggestion
   * @nodoc
```

AndrewKushnir reviewed (commented):
@yharaskrik thanks for additional changes! 👍
The changes are looking great, I've just added a few more comments.

yharaskrik commented:
> @yharaskrik thanks for additional changes! 👍 The changes are looking great, I've just added a few more comments.

On it!

yharaskrik commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -381,6 +401,35 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
     this.renderer.setAttribute(this.imgElement, name, value);
   }
+
+  private createPreloadLinkTag(url: string, srcset: string, sizes?: string): void {
```
good idea, will change.

yharaskrik commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -381,6 +401,35 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+    preload.setAttribute('as', 'image');
+    preload.href = url;
+    preload.rel = 'preload';
+    if (sizes) {
+      preload.imageSizes = sizes;
+    }
+    preload.imageSrcset = srcset;
```
Good idea! Then no one needs to go through the same learning process as me haah

AndrewKushnir reviewed (commented):
@yharaskrik looks great, just left a few last comments.

@kara could you please take a quick look as well (especially around the `sizes` input, which might possibly interfere with #47547)?

yharaskrik commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -381,6 +402,58 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+   * @param src
+   * @param srcset
+   * @param sizes
+   * @private
```
Sounds good! Webstorm added that automatically when I created the comment, I can remove it!

yharaskrik commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -381,6 +402,58 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+   *
+   * {@link https://web.dev/preload-responsive-images/#imagesrcset-and-imagesizes}
+   *
+   * @param src
+   * @param srcset
+   * @param sizes
```
Yup!

yharaskrik commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -381,6 +402,58 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+   * @param sizes
+   * @private
+   */
+  private createPreloadLinkTag({src, srcset, sizes}:
```
Yup we can! I just had multiple undefined ones here before. I can switch it back no problem!

yharaskrik commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -381,6 +402,58 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+   * @description Add a preload `<link>` to the `<head>` of the `index.html` that is served from the
+   * server while using Angular Universal and SSR to kick of image loads for high priority images.
+   *
+   * The `sizes` (passed in from the user) and `srcset` (parsed and formatted from `ngSrcset`)
```
You are welcome! Thanks for the suggestion I think it will help in the future!

yharaskrik commented:
> @yharaskrik looks great, just left a few last comments.
> 
> @kara could you please take a quick look as well (especially around the `sizes` input, which might possibly interfere with #47547)?

Yes this will definitely conflict with that other PR, if that one is merged first then I can just rebase and merge or take one of the comments but the functionality is the same. (`@Input() sizes?: string`)

AndrewKushnir reviewed (approved):
Thanks again, the changes look great 👍

yharaskrik commented:
> Thanks again, the changes look great 👍

Thank you for the patience!

AndrewKushnir commented:
@yharaskrik we've noticed that there are a number of commits in this PR. All commits will be picked up by the changelog generator and appear in the CHANGELOG.md file. Could you please merge them into one?

yharaskrik commented:
> @yharaskrik we've noticed that there are a number of commits in this PR. All commits will be picked up by the changelog generator and appear in the CHANGELOG.md file. Could you please merge them into one?

Should be good now, let me know if I did something wrong, haven't done that before.

atscott reviewed (approved):
reviewed-for: public-api

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -285,6 +291,13 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+   * This input is exclusively read to pass along to the preload `<link>` attribute for this image.
+   * @nodoc
+   */
+  @Input() sizes?: string;
```
I think you're going to want to set the host attribute for `sizes` (ex: L333 or L346) explicitly to make sure the attribute is mirrored to the HTML element if it's bound. e.g `<img [sizes]="someSizesValue">` won't actually pass through the "sizes" attribute to the img tag by default, so you have to pass it manually. (This is what we had to do with width and height)

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -381,6 +402,56 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+
+  /**
+   * @description Add a preload `<link>` to the `<head>` of the `index.html` that is served from the
+   * server while using Angular Universal and SSR to kick of image loads for high priority images.
```
```suggestion
   * server while using Angular Universal and SSR to kick off image loads for high priority images.
```

kara reviewed (approved):
LGTM, just a few minor things (could be in a follow-up)

dylhunn reviewed (approved):
reviewed-for: public-api

AndrewKushnir commented:
[Presubmit](http://test/OCL:478909373:BASE:478909402:1664925768210:fd0f9de7).

pkozlowski-opensource commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -323,7 +484,7 @@ describe('Image directive', () => {
 
         const template = `
-            <img ngSrc="img" ngSrcset="" width="100" height="50">
+            <img ngSrc='img' ngSrcset='' width='100' height='50'>
```
nit: could you please revert all the `"` => `'` changes and removed comas? This change makes the diff necessary bigger and shouldn't be part of this PR.

yharaskrik replied:
Huh, I have no idea how that happened, I can 100% change those back! Sorry about that.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -381,6 +403,56 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+   * @param srcset The parsed and formatted srcset created from the `ngSrcset` input
+   * @param sizes The value of the `sizes` attribute passed in to the `<img>` tag
+   */
+  private createPreloadLinkTag(src: string, srcset?: string, sizes?: string): void {
```
can we move this function to a dedicated service? I find it a bit strange that we inject `PRELOADED_IMAGES` Set and have the actual logic here. IMO the code flow would benefit from co-locating things in a service.

yharaskrik replied:
Was thinking `class PreloadLinkCreator {}`?

pkozlowski-opensource reviewed (dismissed):
IMO we should clarify https://github.com/angular/angular/pull/47343#pullrequestreview-1131215623

yharaskrik commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -319,14 +332,23 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+    }
+
+    if (this.isServer && this.priority) {
+      this.createPreloadLinkTag(this.ngSrc, rewrittenSrcset, this.sizes);
```
So the link Kara gave me has the `href` of the preload link using the "original" src of the image, now of course the link was not dealing with image loaders or this directive specifically just explaining the imagesrcset and imagesizes attributes on the `link` tag, so once we introduce the `IMAGE_LOADER` maybe the `rewrittenSrc` IS the "original" src in this case?

kara replied:
Oh, good catch. It should definitely be the rewrittenSrc (it's the full constructed URL)

yharaskrik replied:
I will switch it back!

yharaskrik commented:
> IMO we should clarify [#47343 (review)](https://github.com/angular/angular/pull/47343#pullrequestreview-1131215623)

Agreed, just left you a comment, I think I may have taken the suggested changes from @kara too literally (based on what the link was describing) as I was originally passing the `rewrittenSrc` rather than `ngSrc` into the `<link>` tag.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/preload-link-creator.ts:
```diff
@@ -0,0 +1,78 @@
+
+    this.preloadedImages.add(src);
+
+    const preload = this.document.createElement('link');
+    preload.setAttribute('as', 'image');
+    preload.href = src;
+    preload.rel = 'preload';
+
+    if (sizes) {
+      preload.imageSizes = sizes;
+    }
+
+    if (srcset) {
+      preload.imageSrcset = srcset;
+    }
+
+    this.document.head.appendChild(preload);
```
@yharaskrik I think it'd be also great to use the renderer abstraction here to make the code consistent with other logic. Btw, quick question on `as` attribute: does it need to be an attribute and other fields are properties (if so, could you please add a comment to mention that)?

```suggestion
    // injected earlier as:
    // private readonly renderer = inject(Renderer2);
    const renderer = this.renderer;
    const preload = renderer.createElement('link');
    renderer.setAttribute(preload, 'as', 'image');
    renderer.setProperty(preload, 'href', src);
    renderer.setProperty(preload, 'rel', 'preload');

    if (sizes) {
      renderer.setProperty(preload, 'imageSizes', sizes);
    }

    if (srcset) {
      renderer.setProperty(preload, 'imageSrcset', srcset);
    }

    renderer.appendChild(this.document.head, preload);
```

yharaskrik replied:
I am not actually sure why i used attr instead of prop there. I will change it to property

AndrewKushnir replied:
> Is that better than just doing this?:
private readonly renderer = inject(RendererFactory2).createRenderer(null, null);
> Not sure if the params to create renderer are important? (especially cause we are only adding to <head> anyway?

(moving the discussion to this thread :))

The renderer can be overridden on a component level, so it'd be a bit safer to pass it from the directive.

yharaskrik replied:
ok sounds good!

AndrewKushnir reviewed (commented):
@yharaskrik thanks again for addressing the feedback! I've left one minor comment on using the renderer abstraction. Thank you.

yharaskrik commented:
> @yharaskrik thanks again for addressing the feedback! I've left one minor comment on using the renderer abstraction. Thank you.

I can do that but will need to use the RendererFactory to create one as I cannot inject Renderer2 into services.

AndrewKushnir commented:
> > @yharaskrik thanks again for addressing the feedback! I've left one minor comment on using the renderer abstraction. Thank you.
> 
> I can do that but will need to use the RendererFactory to create one as I cannot inject Renderer2 into services.

You are right, in this case we can pass the renderer from the image directive during the `createPreloadLinkTag` call. WDYT?

yharaskrik commented:
> > > @yharaskrik thanks again for addressing the feedback! I've left one minor comment on using the renderer abstraction. Thank you.
> > 
> > 
> > I can do that but will need to use the RendererFactory to create one as I cannot inject Renderer2 into services.
> 
> You are right, in this case we can pass the renderer as an argument from the image directive during the `createPreloadLinkTag` call. WDYT?

Is that better than just doing this?:
`private readonly renderer = inject(RendererFactory2).createRenderer(null, null);`

Not sure if the params to create renderer are important? (especially cause we are only adding to `<head>` anyway?

yharaskrik commented on packages/common/src/directives/ng_optimized_image/preload-link-creator.ts:
```diff
@@ -0,0 +1,79 @@
+    this.preloadedImages.add(src);
+
+    const preload = renderer.createElement('link');
+    renderer.setAttribute(preload, 'as', 'image');
+    renderer.setProperty(preload, 'href', src);
+    renderer.setProperty(preload, 'rel', 'preload');
```
So use set attribute for all 3 then? I do have some tests that ensure the as and the rel are set properly. I think it's the first spec file. Would you like more?

AndrewKushnir replied:
I did some local testing and I think both attribute and property would work - all properties that we've used here are reflected as attributes on an element as well (I wasn't sure about `imageSizes` and `imageSrcset`, but they work too, though the attribute name is lowercased). I think we could still update this code to use `setAttribute`, so that it's more explicit that we want all values to appear as attributes on a `<link>` during serialization. I see that you have tests that verify that "as" and "rel" can be accessed via "getAttribute", we can add similar checks for the remaining fields as well.

yharaskrik replied:
Sounds good. I'll get that done tomorrow. And they are lowercase because that's what the TS typings have them as so when I switched it to setProperty I used the lowercase. I don't think it matters though aren't attributes not case sensitive?

yharaskrik replied:
setAttributes works and the attributes are case insensitive like i thought they were. You can see that in the tests.

AndrewKushnir commented:
@yharaskrik the PR #47547 that adds `sizes` input as well has landed. Could you please rebase this PR, so that we can complete the final round of code reviews and proceed with further testing in Google's codebase? Thank you.

yharaskrik commented:
There seems to be some seemingly unrelated test failures now and I cannot for the life of me track them down. It's like there are some leaks between tests as some tests are getting entirely different values than are configured on the `<img>` tag.

AndrewKushnir commented:
> There seems to be some seemingly unrelated test failures now and I cannot for the life of me track them down. It's like there are some leaks between tests as some tests are getting entirely different values than are configured on the <img> tag.

@yharaskrik thanks for highlighting this issue, I will look into this now and let you know.

AndrewKushnir committed:
fixup! feat(common): add <link> preload tag on server for priority img

AndrewKushnir commented:
@yharaskrik it looks like there was an issue with the rebase and an extra condition that was added in the PR https://github.com/angular/angular/pull/47547 got inserted in a different place. I've added a fixup commit to resolve this: https://github.com/angular/angular/pull/47343/commits/2969b99a3150dd1f085cdca653de694f9989374b

atcastle commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -8,16 +8,186 @@
+      });
+
+      const template =
+          `<img ngSrc="${src}" width="150" height="50" priority sizes="10vw" ngSrcset="100w">`;
```
Let's also include a test with no ngSrcset to make sure that the link element is picking up the automatically-generated srcset

yharaskrik replied:
Ok I'll try and get to that today

AndrewKushnir replied:
@yharaskrik let's address this feedback in a followup PR (adding changes to this PR would reset approvals).

yharaskrik replied:
👌

atcastle reviewed (approved):
This looks good to me--I checked where the merge conflicts were resolved with the automatic srcset PR and it seems like it'll work fine.

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:478909373:BASE:480149161:1665429714265:2f9f681f) (internal link).

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
@yharaskrik FYI, the current state of this PR is "ready for merge", I'm adding it to the merge queue. There is a feedback from @atcastle (see https://github.com/angular/angular/pull/47343#discussion_r991577560), which we can handle in a followup PR. Please do not push any additional changes for now (it'd reset all approvals). I will let you know once this PR is merged.

AndrewKushnir commented:
Merge-assistance (@jessicajaniuk): this PR is ready for merge. Pawel's feedback has been addressed, but since he is OOO, we can't change the status.

yharaskrik commented:
> @yharaskrik FYI, the current state of this PR is "ready for merge", I'm adding it to the merge queue. There is a feedback from @atcastle (see https://github.com/angular/angular/pull/47343#discussion_r991577560), which we can handle in a followup PR. Please do not push any additional changes for now (it'd reset all approvals). I will let you know once this PR is merged.

Ok sounds good!

jessicajaniuk commented:
This PR was merged into the repository by commit 75e6297f0901cc98aea1626a138a820e68d026ec.

AndrewKushnir commented:
@yharaskrik just wanted to say **thank you** once again for all your work on this PR 👍
The PR is now merged and the change will be released as a part of the upcoming Angular v15.

yharaskrik commented:
> @yharaskrik just wanted to say **thank you** once again for all your work on this PR 👍
> The PR is now merged and the change will be released as a part of the upcoming Angular v15.

Thank you for all the help and guidance! Hope I can contribute again soon! I'd be curious what the future plans of the image directive is as I think that's a great step forward! (I'm in the angular slack as well)
